synapse/handlers/auth.py

```diff
@@ -754,13 +756,48 @@ def _auth_dict_for_flows(
             "params": params,
         }
 
+    async def refresh_token(
+        self, refresh_token: str, valid_until_ms: Optional[int],
+    ) -> Tuple[str, str]:
+        existing_token = await self.store.lookup_refresh_token(refresh_token)
+        if existing_token is None:
+            raise SynapseError(400, "refresh token does not exist")
+
+        if not existing_token.valid:
+            raise SynapseError(400, "refresh token isn't valid anymore")
+
+        (
+            new_refresh_token,
+            new_refresh_token_id,
+        ) = await self.get_refresh_token_for_user_id(
+            user_id=existing_token.user_id, device_id=existing_token.device_id
+        )
+        access_token = await self.get_access_token_for_user_id(
+            user_id=existing_token.user_id,
+            device_id=existing_token.device_id,
+            valid_until_ms=valid_until_ms,
+            refresh_token_id=new_refresh_token_id,
+        )
+        await self.store.invalidate_refresh_token(existing_token.token_id)
```

Comment:
We'll need to make this API idempotent somehow, i.e. if the client never gets the response to their `/refresh` request they need to be able to retry it without failing. One possible option would be to not invalidate the refresh token until the server sees the new access token in use, but there are probably other ways of doing so.

Reply:
I'm now doing this by checking if any "child" refresh token was already used. I consider that a refresh token has been used either if the associated access token was used (using the `last_validated` field) or if the refresh token was consumed (used on `/refresh`).
I'll write try to write some tests to check if the behavior is correct.